File: src/funciones/funciones.py
from typing import Callable
#Función1

def producto(n:int, k:int) -> int:
    n, k = max(n, k), min(n, k)
    s:int = 1
    for i in range(k+1):
        s*= n- i + 1
        
    return s



#Función2
def secuencia(a1:int, r:int, k:int) -> int:
    ls: list[int] = [a1]
    for i in range(2, k+1):
        ls.append(a1*r**(i-1))
    s:int = 1
    for n in ls:
        s*=n
    
    return s



#Función3
def combinatorio(n:int, k:int) -> int:
    n, k = max(n, k), min(n, k)
    def factorial(t):
        r:int = 1
        for i in range(t, 1, -1):
            r*=i
        return r
    
    s:int = factorial(n)/(factorial(k)*factorial(n-k))
    
    return s



#Función4
def numeroS(n, k) -> float:
    n, k = max(n, k), min(n, k)
    def factorial(t):
        r:int = 1
        for i in range(t, 1, -1):
            r*=i
        return r
    
    s:int = 1/factorial(k)
    t:int = 0
    for i in range(k):
        t += (-1)**i * combinatorio(k+1, i+1) * (k-i)**n
        
    return s*t
    


#Función5
def Newton(a:float, e:float, f:Callable[[float], float], df:Callable[[float], float]) -> float:
    x0:float = a
    funcion:float = float(f(x0))
    while abs(funcion) > e:
        x0 -= f(x0)/df(x0)
        funcion = float(f(x0))
    
    return x0




    
    

File: src/Entrega1/Funciones.java
package Entrega1;

import java.util.function.Function;

public class Funciones {
	
	public static Integer Factorial(Integer n) {
		Integer r = 1;
		for(int i=n; i>=1; i--) {
			r *= i;
		}
		return r;
	}
	
	public static Integer Producto(Integer n, Integer k) {
		assert n>k : String.format("n debe ser mayor que k");
		Integer p = 1;
		for(int i=0; i<=k; i++) {
			p *= n-i+1;		
		}
		return p;	
	}
		
	public static Integer Secuencia(Integer a, Integer r, Integer k) {
		Integer result= 1;
		for (int i=1; i<=k; i++) {
			result *= a * (int)Math.pow(r, i-1);
		}
		
		return result;
	}
	
	public static Integer Combinatorio(Integer n, Integer k) {
		assert n>=k : String.format("n debe ser mayor o igual que k");
		return Factorial(n)/(Factorial(n-k)*Factorial(k));
	}
	
	public static Double S(Integer n, Integer k) {
		assert n>=k : String.format("n debe ser mayor o igual que k");
		Double s = 0.;
		for(int i=0; i<=k-1; i++) {
			s += Math.pow(-1, i) * Combinatorio(k+1, i+1) * Math.pow(k-i, n);
		}
		return (1./Factorial(k))*s;
	}
	
	public static Double Newton(Function<Double, Double> f, Function<Double, Double> df, Double a, Double e) {
		Double valor = a;
		while(! (Math.abs(f.apply(valor))<= e)) {
			valor -= f.apply(valor)/df.apply(valor);
		}		
		return valor;		
	}

	public static void main(String[] args) {
		System.out.println(Factorial(3));
		System.out.println(Producto(4, 2));
		System.out.println(Secuencia(3, 5, 2));
		System.out.println(Combinatorio(4, 2));
		System.out.println(S(4, 2));
		System.out.println(Newton(x -> 2*Math.pow(x, 2), x -> 4*x, 3., 0.001));

	}

}
